ci_env.sh

```diff
@@ -30,4 +30,9 @@ if [[ $HIGHEST_CHAIN_TAG == "mainnet" ]]; then
     echo "disallowing mainnet release without semver tag; $generatedVersion != $definedVersion"
     exit 1
   fi
+else
+  if [[ $generatedVersion == $definedVersion ]]; then
```

Comment:
So the CircleCI tests won't run in the case when `mainnet` is merged into `rinkeby` (or `rinkeby` into `master`) after a version bump since this would evaluate to true and the `build-builder` job in CircleCI would fail preventing the `build` step (with the tests) from running?

Reply:
Is there a particular reason we don't make the release commit on dev? Presumably this is just bumping the versions file.

If we can fast-forward `master` merges into `mainnet` then the resulting SHA will end up the same. (But if not, it's not a big deal.) The tag can then be created on mainnet branch - the tag doesn't really need to be ported back into `dev`.

Reply:
Good point, I'll write it up that way instead.

Reply:
Wait, I remember why — it's because tags don't point to a particular branch like that. Say I run all these commands at once:

```bash
git checkout master
echo -n '0.5.2' > VERSION
git commit -am 'release v0.5.2'
git push

git checkout mainnet
git merge --ff-only master
git tag v0.5.2
git push --atomic origin mainnet v0.5.2
```

And only after that does the CircleCI/Travis/Azure build start. The build for `master` will start and notice that its current commit has a tag and treat that as the version number.

I suppose I could fix this by having `print_version.sh` only output the tag commit without its SHA if you're on `mainnet`, but that would break building tagged versions locally. So I dunno. Happy to consider alternatives though.

Reply:
Could we incorporate the branch into the version tag? Something like `v0.5.2-mainnet` or `v0.5.2-rinkeby` ? Then builds would only occur for that particular branch?

Reply:
> Could we incorporate the branch into the version tag? Something like v0.5.2-mainnet or v0.5.2-rinkeby ?

Adding on to this, what if we updated `print_version.sh` to something like this:

- If the `currentTag` = `currentVersion` and the branch is not `mainnet`, print `currentVersion-<BRANCH>`
- If the `currentTag` = `currentVersion` and the branch is `mainnet`, print `currentVersion`
- If the `currentTag` != `currentVersion` and the branch is not `mainnet`, print `currentVersion-<BRANCH>-<SHA>` [1]
- If the `currentTag` != `currentVersion` and the branch is `mainnet`, print `currentVersion-<SHA>` [2]

[1] Might be helpful to differentiate non-release CI builds created from `master` vs. `rinkeby` if we merge `master` into `rinkeby` periodically to do testing prior to a release.

[2] Technically, we would never create a build for this case since we have the rule that requires a semver tag for mainnet releases so we don't really need to handle this case, but adding it for completeness.

After the `v0.5.2` commit is created on `master, if we merge `master` -> `rinkeby` -> `mainnet` we would end up with the following version tags for each branch:

`master` -> `v0.5.2-master`
`rinkeby` -> `v0.5.2-rinkeby`
`mainnet` -> `v0.5.2`

Reply:
@j0sh 

> Could we incorporate the branch into the version tag? Something like v0.5.2-mainnet or v0.5.2-rinkeby ? Then builds would only occur for that particular branch?

As actual git tags? It seems to me that `v0.5.2` should uniquely identify the state of the code at a given time; having both `v0.5.2-rinkeby` and `v0.5.2-mainnet` muddles that reality somewhat.

@yondonfu  I'm not sure `print_version.sh` should be branch-aware. I'm concerned about the use case of a user building themselves a binary locally.

```bash
git checkout v0.5.2
HIGHEST_BUILD_TAG=mainnet make livepeer
```

That should deliver you a mainnet-compatible binary that identifies itself as v0.5.2. That conflicts with your first rule above (If the `currentTag` = `currentVersion` and the branch is not `mainnet`, print `currentVersion-<BRANCH>`).

Reply:
But maybe that just means we treat `mainnet` and the tag-but-no-branch case identically. That makes your rules work okay, I think.

Reply:
> I'm concerned about the use case of a user building themselves a binary locally.
> That should deliver you a mainnet-compatible binary that identifies itself as v0.5.2.

Ah ok I see what you're saying now.

If we're merging a PR with the release commit to `master` and then fast-forward merging `master` into `rinkeby` and `master`, perhaps its ok for CI to just not build anything for `master` and `rinkeby` since the binary built for `mainnet` would be functionally identical to the binaries that would be built for `master` and `rinkeby`. Furthermore, the binary built for `mainnet` would be usable for any use case that one might've had for a binary built for `master` and/or `rinkeby` since the `mainnet` binary will support Rinkeby as well as any dev network. Only building a single binary when there is a semver tag release also feels simpler. 

I would be a proponent of:
- Keeping this check to disallow semver tag release builds on any branch besides `mainnet`
- Making the release process:

1. Make release commit and push semver release tag on a `vX.Y.Z` branch
2. Create PR merging `vX.Y.Z` into `master` and merge PR into `master`. This will not create a binary
3. Fast forward merge `master` into `rinkeby`. This will not create a binary
4. Fast forward merge `master` into `mainnet`. This will crease a binary that identifies itself as `vX.Y.Z`

Reply:
PR with suggested release process #1329